### PR TITLE
kolla: init at 18.1.0

### PR DESCRIPTION
**Pull request by astro**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ko/kolla/package.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+, python311
+, fetchgit
+, bashate
+}:
+
+let
+  pythonPackages = python311.pkgs;
+
+in
+pythonPackages.buildPythonPackage rec {
```
Comment (mweinelt) [inline pkgs/by-name/ko/kolla/package.nix:11]:
```suggestion
, python311Packages
, fetchgit
, bashate
}:

python311Packages.buildPythonApplication rec {
```

`buildPythonPackage` is for libraries, you'll probably want this. Also, you can consume `python311Packages` directly.

File: pkgs/by-name/ko/kolla/package.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      "os.path.join(PROJECT_ROOT, '../../../share/kolla'),"
+
+    substituteInPlace test-requirements.txt \
+      --replace "hacking>=3.0.1,<3.1.0" "hacking"
+
+    sed -e 's/git_info = .*/git_info = "${version}"/' -i kolla/version.py
+  '';
+
+  # fake version to make pbr.packaging happy
+  PBR_VERSION = version;
```
Comment (mweinelt) [inline pkgs/by-name/ko/kolla/package.nix:35]:
```suggestion
  # fake version to make pbr.packaging happy
  env.PBR_VERSION = version;
```

We try to leak things explicitly now.

File: pkgs/by-name/ko/kolla/package.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  propagatedBuildInputs = with pythonPackages; [
+    pbr
+    jinja2
+    oslo-config
+    gitpython
+    pythonPackages.docker
+    pythonPackages.podman
+  ];
```
Comment (mweinelt) [inline pkgs/by-name/ko/kolla/package.nix:44]:
```suggestion

  build-system = with pythonPackages; [
    setuptools
    pbr
  ];

  dependencies = with pythonPackages; [
    jinja2
    oslo-config
    gitpython
    pythonPackages.docker
    pythonPackages.podman
  ];
```

Separate build and runtime dependencies into `build-system` and `dependencies`, which is the PEP517 lingo for `nativeBuildInputs` and `propagatedBuildInputs`.

File: pkgs/by-name/ko/kolla/package.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    rev = version;
+  };
+
+  postPatch = ''
+    substituteInPlace kolla/image/kolla_worker.py \
+      --replace "os.path.join(sys.prefix, 'share/kolla')," \
+      "os.path.join(PROJECT_ROOT, '../../../share/kolla'),"
+
+    substituteInPlace test-requirements.txt \
+      --replace "hacking>=3.0.1,<3.1.0" "hacking"
```
Comment (SuperSandro2000) [inline pkgs/by-name/ko/kolla/package.nix:29]:
```suggestion
      --replace-fail "os.path.join(sys.prefix, 'share/kolla')," \
      "os.path.join(PROJECT_ROOT, '../../../share/kolla'),"

    substituteInPlace test-requirements.txt \
      --replace-fail "hacking>=3.0.1,<3.1.0" "hacking"
```

File: pkgs/by-name/ko/kolla/package.nix [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  pname = "kolla";
+  version = "18.1.0";
+
+  pyproject = true;
+
+  src = fetchgit {
+    url = "https://opendev.org/openstack/kolla";
+    hash = "sha256-jLD6ILihymQlWkkpGYC4OX8BKLpQurAK6Y5Xpju+QAI=";
+    rev = version;
+  };
```
Comment (mweinelt) [inline pkgs/by-name/ko/kolla/package.nix:21]: Can we use fetchGitea here? They look compatible at first glance. Fetching a tarball would be a lot cheaper.

Comment (astro) [reply pkgs/by-name/ko/kolla/package.nix:21]: Unfortunately, their Gitea is somewhat broken:
```
--2024-08-06 15:29:45--  https://opendev.org/openstack/kolla/archive/18.1.0.tar.gz
Resolving opendev.org (opendev.org)... 38.108.68.66
Connecting to opendev.org (opendev.org)|38.108.68.66|:443... connected.
HTTP request sent, awaiting response... No data received.
```

There's also an automated mirror on Github. Do you prefer that or should we stick with `fetchgit` from the authoritative source?

Comment (SuperSandro2000) [reply pkgs/by-name/ko/kolla/package.nix:21]: I think in other places I have used the github mirror.

File: pkgs/by-name/ko/kolla/package.nix [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    setuptools
+    pbr
+  ];
+
+  dependencies = with pythonPackages; [
+    jinja2
+    oslo-config
+    gitpython
+    pythonPackages.docker
+    pythonPackages.podman
```
Comment (SuperSandro2000) [inline pkgs/by-name/ko/kolla/package.nix:47]:
```suggestion
    docker
    jinja2
    oslo-config
    gitpython
    podman
```

**Comment by astro** [comment]
But I *did* run nixfmt.

**Comment by SuperSandro2000** [comment]
nixfmt-classic or nixfmt-rfc-style? 😅

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**